taf/tests/test_updater/test_update_library/test_validation_and_sync.py

```diff
@@ -0,0 +1,294 @@
+    "origin_auth_repo",
+    [
+        {
+            "targets_config": [
+                {"name": "target1", "allow_unauthenticated_commits": True},
+                {"name": "target2", "allow_unauthenticated_commits": True},
+            ],
+        },
+    ],
+    indirect=True,
+)
+def test_target_repo_not_in_sync_partial(origin_auth_repo, client_dir):
+    """
+    Same as 2, but target repositories contain invalid commits following commits which were manually pulled using git pull.
+    Expect all repositories to be updated up to the last validated commits and invalid commits to be removed
+    """
+    clone_repositories(
+        origin_auth_repo,
+        client_dir,
+    )
+    target_repos = list(load_target_repositories(origin_auth_repo).values())
+    original_commits = {repo.name: repo.head_commit_sha() for repo in target_repos}
+
+    setup_manager = SetupManager(origin_auth_repo)
+    setup_manager.add_task(add_valid_unauthenticated_commits)
+    setup_manager.execute_tasks()
+
+    new_commits = {repo.name: repo.head_commit_sha() for repo in target_repos}
+    for repo_name in original_commits:
+        assert original_commits[repo_name] != new_commits[repo_name]
+    setup_manager.add_task(
+        sync_target_repos_with_remote,
+        kwargs={"origin_auth_repo": origin_auth_repo, "client_dir": client_dir},
+    )
+    setup_manager.execute_tasks()
+
+    update_and_check_commit_shas(OperationType.UPDATE, origin_auth_repo, client_dir)
+    check_repo_sync(client_dir, origin_auth_repo)
+
+
+@pytest.mark.parametrize(
+    "origin_auth_repo",
+    [
+        {
+            "targets_config": [
+                {"name": "target1", "allow_unauthenticated_commits": True},
+                {"name": "target2", "allow_unauthenticated_commits": True},
+            ],
+        },
+    ],
+    indirect=True,
+)
+def test_mixed_target_repo_states(origin_auth_repo, client_dir):
+    """
+    Client's auth repo is on last validated commit, but target commits are all over the place.
+    Some are after last validated, some are before.
+    """
+    target_repos = list(load_target_repositories(origin_auth_repo).values())
+    original_commits = {repo.name: repo.head_commit_sha() for repo in target_repos}
+    reverted_repo = target_repos[0]  # target1
+    updated_repo = target_repos[1]  # target2
+
+    clone_repositories(
+        origin_auth_repo,
+        client_dir,
+    )
+
+    # Add Valid commits so target1 can be reverted
+    setup_manager = SetupManager(origin_auth_repo)
+    setup_manager.add_task(add_valid_target_commits)
+    setup_manager.add_task(update_expiration_dates)
+    setup_manager.execute_tasks()
+
+    update_and_check_commit_shas(OperationType.UPDATE, origin_auth_repo, client_dir)
+
+    # revert target1, add to target2
+    setup_manager.add_task(
+        remove_commits, kwargs={"repo_path": reverted_repo.path, "num_commits": 1}
+    )
+    setup_manager.add_task(
+        update_target_files,
```

Comment:
I think that this just updates the auth repo's target files and not the actual target repo. We could just manually pull again (but only one target repo)